src/lock.rs

```diff
@@ -0,0 +1,328 @@
+///
+/// It is used to ensure that future installations will use the exact same dependencies.
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LockedPackage {
+    pub name: PackageName,
+    pub digest: Digest,
+    pub registry: RegistryUri,
+    pub repository: String,
+    pub version: Version,
+    pub dependencies: Vec<PackageName>,
+    pub dependants: usize,
+}
+
+impl LockedPackage {
+    /// Captures the source, version and checksum of a Package for use in reproducible installs
+    pub fn lock(
+        package: &Package,
+        registry: RegistryUri,
+        repository: String,
+        dependants: usize,
+    ) -> Self {
+        let digest = digest::digest(&digest::SHA256, &package.tgz)
+            .try_into()
+            .unwrap(); // won't fail as SHA256 is sure to be supported
+
+        Self {
+            name: package.name().to_owned(),
+            registry,
+            repository,
+            digest,
+            version: package.version().to_owned(),
+            dependencies: package
+                .manifest
+                .dependencies
+                .iter()
+                .cloned()
+                .map(|d| d.package)
+                .collect(),
+            dependants,
+        }
+    }
+
+    /// Validates if another LockedPackage matches this one
+    pub fn validate(&self, package: &Package) -> eyre::Result<()> {
+        let digest: Digest = digest::digest(&digest::SHA256, &package.tgz)
+            .try_into()
+            .unwrap();
+
+        ensure!(
+            &self.name == package.name(),
+            "Package name mismatch - expected {}, actual {}",
+            self.name,
+            package.name()
+        );
+        ensure!(
+            self.digest == digest,
+            "Digest mismatch - expected {}, actual {}",
+            self.digest,
+            digest
+        );
+        ensure!(
+            &self.version == package.version(),
+            "Version mismatch - expected {}, actual {}",
+            self.version,
+            package.version()
+        );
+
+        Ok(())
+    }
+
+    /// Constructs a Dependency instance with matching metadata
+    pub fn as_dependency(&self) -> Dependency {
+        Dependency {
+            package: self.name.clone(),
+            manifest: DependencyManifest {
+                version: VersionReq {
+                    comparators: vec![semver::Comparator {
+                        op: semver::Op::Exact,
+                        major: self.version.major,
+                        minor: Some(self.version.minor),
+                        patch: Some(self.version.patch),
+                        pre: self.version.pre.clone(),
+                    }],
+                },
+                registry: self.registry.clone(),
+                repository: self.repository.clone(),
+            },
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct RawLockfile {
```

Comment:
Given that this is versioned, I would probably write it as an `enum` straight away so that you don't need to change lots of code later when you _do_ cut a new version.

FYI, you'll probably want to write your own deserializer to switch on `version` is my guess.

Reply:
You don't _have_ to do it in this PR (it's already quite large), but probably worth doing proactively.

Reply:
Hmm if we're writing our own deserializer, is there a point to having `RawLockfile` at all? It's only there as an intermediate format that can be directly serialized/deserialized with the default toml serde implementation.

I think I'll keep it as is for now since this is a private struct and I don't expect anything will break _outside this module_ if we change how ser/de works later, but correct me if I'm wrong.